### PR TITLE
Document the core rules

### PR DESCRIPTION
**Pull request by ianthehat**
More for #810

Still need the x_ref section filled in, and some more about embedding and why we do it, but even without that it is strictly better and more accurate than what we have right now.

[skip ci]
[ci skip]

### CODE REVIEW

File: go/core.rst [hunk 1]
```diff
@@ -0,0 +1,425 @@
...
+
+Defines and stamping
+~~~~~~~~~~~~~~~~~~~~
+
+**TODO**: More information
+
+Embedding
+~~~~~~~~~
+
+This is used for things like internal tests, where a library is recompiled with additional
```
Comment (jayconrod) [inline go/core.rst]: Missing word after "additional".

Also s/dependancies/dependencies/g

File: go/core.rst [hunk 2]
```diff
@@ -0,0 +1,425 @@
...
+Providers
+^^^^^^^^^
+
+* GoLibrary_
+* GoEmbed_
+
+Output groups
+^^^^^^^^^^^^^
+
+* default *: A library with the default build options.*
```
Comment (jayconrod) [inline go/core.rst]: Would be better to have the group names emphasized and the descriptions in normal face.

File: go/core.rst [hunk 3]
```diff
@@ -0,0 +1,425 @@
...
+| :param:`srcs`              | :type:`label_list`          | :value:`None`                         |
++----------------------------+-----------------------------+---------------------------------------+
+| The list of Go source files that are compiled to create the package.                             |
+| Only :value:`.go` files are permitted, unless the cgo attribute is set, in which case the        |
+| following file types are permitted: :value:`.go, .c, .s, .S .h`.                                 |
+| The files may contain Go-style `build constraints`_.                                             |
++----------------------------+-----------------------------+---------------------------------------+
+| :param:`deps`              | :type:`label_list`          | :value:`None`                         |
++----------------------------+-----------------------------+---------------------------------------+
+| List of GoLibrary_ providers this library imports directly.                                      |
```
Comment (jayconrod) [inline go/core.rst]: I would rephrase this as "List of Go libraries this library imports directly. These may be `go_library` rules or compatible rules with the `GoLibrary` provider."

File: go/core.rst [hunk 4]
```diff
@@ -0,0 +1,425 @@
...
++----------------------------+-----------------------------+---------------------------------------+
+| If :value:`True`, the package uses cgo_.                                                         |
+| The cgo tool permits Go code to call C code and vice-versa.                                      |
+| This does not support calling C++.                                                               |
+| When cgo is set, :param:`srcs` may contain C or assembly files; these files are compiled with    |
+| the normal c compiler and included in the package.                                               |
++----------------------------+-----------------------------+---------------------------------------+
+| :param:`cdeps`             | :type:`label_list`          | :value:`None`                         |
++----------------------------+-----------------------------+---------------------------------------+
+| The list of other libraries that the c code depends on.                                          |
```
Comment (jayconrod) [inline go/core.rst]: List of `cc_library` rules the cgo code depends on. Only valid if `cgo = True`.

(Though I don't think we actually check that).

File: go/core.rst (hunk 2 above)
Comment (ianthehat) [reply go/core.rst]: Done

File: go/core.rst (hunk 1 above)
Comment (ianthehat) [reply go/core.rst]: Done

File: go/core.rst [hunk 5]
```diff
@@ -0,0 +1,425 @@
...
+
+.. role:: param(kbd)
+.. role:: type(emphasis)
+.. role:: value(code)
+.. |mandatory| replace:: **mandatory value**
+
+These are the core go rules, required for basic operation.
+The intent is that theses rules are sufficient to match the capabilities of the normal go tools.
+
+.. contents::
```
[the remark this answers is not in the document]
Comment (ianthehat) [reply go/core.rst]: Yup, done.

File: go/core.rst (hunk 3 above)
Comment (ianthehat) [reply go/core.rst]: Done

File: go/core.rst (hunk 4 above)
Comment (ianthehat) [reply go/core.rst]: Done
